In [1]:
#!pip install spacy
#!pip install spacy en_core_web_sm

#import nltk
#nltk.download('punkt')

import pandas as pd
import numpy as np
import re
#import spacy
#from spacy import displacy

pd.set_option('display.max_columns', None)  # me muestre todas las columnas
pd.set_option('display.max_colwidth', 100)  #me muestre más caracteres por columna.

import warnings
warnings.filterwarnings('ignore')

# para pintar
import pylab as plt
import seaborn as sns

# para que salga el grafico
%matplotlib inline

In [2]:
data = pd.read_csv('attacks.csv', encoding= "ISO-8859-1") 

In [3]:
data.info(memory_usage = 'deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

In [4]:
data[data['Unnamed: 23'] == 'Teramo']           

Case Number         Date    Year        Type               Country  \
4415  1952.03.30  30-Mar-1952  1952.0  Unprovoked  NETHERLANDS ANTILLES   

         Area Location                                         Activity  \
4415  Curacao      NaN  Went to aid of child being menaced by the shark   

             Name Sex   Age                          Injury Fatal (Y/N) Time  \
4415  A.J. Eggink    M  NaN  Buttock bitten, tissue removed           N  NaN   

                                                                                                 Species   \
4415  Bull shark, 2.7 m [9'] was captured & dragged on the sand where tissue taken from Eggink was fou...   

                                             Investigator or Source  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. Baldridge, p.172   

                        pdf  \
4415  1952.03.30-Eggink.pdf   

                                                                     href formula  \
4415  http://sharkattackfile.net/spreadsheets/pdf_directory/1952.03.30-Eggink.pdf   

                                                                             href  \
4415  http://sharkattackfile.net/spreadsheets/pdf_directory/1952.03.30-Eggink.pdf   

     Case Number.1 Case Number.2  original order Unnamed: 22 Unnamed: 23  
4415    1952.03.30    1952.03.30          1888.0         NaN      Teramo

In [5]:
data[data['Unnamed: 23'] == 'change filename']

Case Number                  Date    Year      Type Country  \
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0  Provoked     USA   

             Area                    Location Activity              Name Sex   \
5840  Connecticut  Branford, New Haven County  Fishing  Captain Pattison    M   

      Age                                        Injury Fatal (Y/N) Time  \
5840  NaN  Leg bitten by netted shark PROVOKED INCIDENT           N  NaN   

     Species         Investigator or Source                        pdf  \
5840      NaN  St. Joseph Herald, 9/14/1878  1878.09.14.R-Pattison.pdf   

                                                                         href formula  \
5840  http://sharkattackfile.net/spreadsheets/pdf_directory/1878.09.14.R-Pattison.pdf   

                                                                                 href  \
5840  http://sharkattackfile.net/spreadsheets/pdf_directory/1878.09.14.R-Pattison.pdf   

     Case Number.1 Case Number.2  original order Unnamed: 22      Unnamed: 23  
5840  1878.09.14.R  1878.09.14.R           463.0         NaN  change filename

In [6]:
data['Unnamed: 23'].value_counts()

Teramo             1
change filename    1
Name: Unnamed: 23, dtype: int64

In [7]:
data['Case Number'].value_counts()

0               2400
1920.00.00.b       2
1966.12.26         2
2014.08.02         2
1990.05.10         2
                ... 
1999.09.05         1
1999.09.10         1
1999.09.16         1
1999.09.18         1
xx                 1
Name: Case Number, Length: 6287, dtype: int64

# Glosario de columnas y tipos iniciales

0. Case Number: La fecha coincide con el case number. OBJETC
1. Date: La fecha del ataque. OBJECT
2. Year: El año del ataque. FLOAT64
3. Type: Tipo de ataque (provocado, no provocado, desastre marino, etc). OBJECT
4. Country: País del ataque. OBJECT
5. Area: Area del país. OBJECT
6. Location: Localización concreta dentro del area. OBJECT
7. Activity: La actividad que se estaba realizando durante el ataque. OBJECT
8. Name: Nombre de la víctima. OBJECT
9. Sex : Género de la víctima. OBJECT
10. Age: Edad de la víctima. OBJECT
11. Injury: Tipo de lesión provocada. OBJECT
12. Fatal (Y/N): Muerte o no de la victima. OBJECT
13. Time: Hora del ataque. OBJECT
14. Species : Especie del tiburón del ataque. OBJECT
15. Investigator or Source: Investigador del ataque + organización a la que pertenece. OBJECT
16. pdf: Imagino que un pdf del informe del ataque por víctima. OBJECT
17. href formula: Enlace al informe en pdf del apartado anterior. OBJECT
18. href: Parece una columna con lo mismo que lo anterior pero tiene menos filas. OBJECT
19. Case Number.1: Columna igual que Case Number pero con dos filas menos. OBJECT
20. Case Number.2: Columna igual que Case Number pero con una fila menos. OBJECT
21. Original Order: Parece un id del caso, a priori mayor cuanto más reciente. FLOAT64
22. Unnamed: 22: No sé qué significa, todo NaNs salvo el 1478 "stopped here". OBJECT
23. Unnamed: 23: No sé qué significa, todo NaNs salvo el 4415 "Teramo" y el 5840 "change filename". OBJECT

# Restricciones:
- No se pueden eliminar columnas.
- Deben quedar al menos 1500 filas.

# Primeras consideraciones

- Como no podemos eliminar columnas, el % de nulos por fila no es tan relevante.
- Sí podemos quitar todas las filas en las que todos los valores sean nulos. 
- Sí podemos quitar todas las filas duplicadas.
- Arreglar los nombres de las columnas para trabajar mejor con ellas.


In [8]:
data_ori_shape = data.shape

data_ori = data.copy()

data.shape

(25723, 24)

In [9]:
data = data.dropna(how='all')

In [10]:
data.shape, data_ori.shape

((8703, 24), (25723, 24))

In [11]:
data= data.drop_duplicates()

In [12]:
data.shape

(6311, 24)

Tras hacerlo, las dimensiones pasan a ser: (6311, 24)

Ahora voy a cambiar los nombres de las columnas para homogeneizarlas, por si acaso.

In [13]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [14]:
data.columns = [e.replace(' ', '_') for e in data.columns]

data.columns = [e.replace('.', '_') for e in data.columns]

data.columns = [e.replace(':', '') for e in data.columns]

In [15]:
data.rename(columns={'Species_': 'Species', 'Sex_': 'Sex', 'Fatal_(Y/N)': 'Fatal'}, inplace=True)

In [16]:
num_unique_values = data.Type.nunique()

En la celda anterior he descubierto que las últimas 9 filas solo tienen un valor no nulo (original_order), así que las elimino.

In [17]:
num_unique_values

8

In [18]:
data['href'].value_counts()

http://sharkattackfile.net/spreadsheets/pdf_directory/w014.01.25-Grant.pdf                 4
http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.02.b-Vandenberg.pdf          3
http://sharkattackfile.net/spreadsheets/pdf_directory/1923.00.00.a-NJ fisherman.pdf        2
http://sharkattackfile.net/spreadsheets/pdf_directory/1934.12.23.a-b-Inman.pdf             2
http://sharkattackfile.net/spreadsheets/pdf_directory/1929.03.04.a-b.Roads-Aldridge.pdf    2
                                                                                          ..
http://sharkattackfile.net/spreadsheets/pdf_directory/1999.09.10-Warnock.pdf               1
http://sharkattackfile.net/spreadsheets/pdf_directory/1999.09.16-Ferguson.pdf              1
http://sharkattackfile.net/spreadsheets/pdf_directory/1999.09.18-Wallersheim.pdf           1
http://sharkattackfile.net/spreadsheets/pdf_directory/1999.09.24-boat.pdf                  1
http://sharkattackfile.net/spreadsheets/pdf_directoryND-0001-Ceylon.pd

In [19]:
print(data.index.is_unique)

True


In [20]:
data['Year'].iloc[15:]

15       2018.0
16       2018.0
17       2018.0
18       2018.0
19       2018.0
          ...  
6306        NaN
6307        NaN
6308        NaN
6309        NaN
25722       NaN
Name: Year, Length: 6296, dtype: float64

In [21]:
len(data['Species'].unique())

1550

In [22]:
data['Year']

0        2018.0
1        2018.0
2        2018.0
3        2018.0
4        2018.0
          ...  
6306        NaN
6307        NaN
6308        NaN
6309        NaN
25722       NaN
Name: Year, Length: 6311, dtype: float64

In [23]:
print(data.iloc[-10:])

      Case_Number       Date  Year        Type             Country  \
6301      ND.0001  1845-1853   0.0  Unprovoked  CEYLON (SRI LANKA)   
6302            0        NaN   NaN         NaN                 NaN   
6303            0        NaN   NaN         NaN                 NaN   
6304            0        NaN   NaN         NaN                 NaN   
6305            0        NaN   NaN         NaN                 NaN   
6306            0        NaN   NaN         NaN                 NaN   
6307            0        NaN   NaN         NaN                 NaN   
6308            0        NaN   NaN         NaN                 NaN   
6309            0        NaN   NaN         NaN                 NaN   
25722          xx        NaN   NaN         NaN                 NaN   

                   Area                             Location  Activity  Name  \
6301   Eastern Province  Below the English fort, Trincomalee  Swimming  male   
6302                NaN                                  NaN       Na

In [24]:
print(data['Case_Number_2'].is_monotonic_increasing)

False


In [25]:
data['Date'].value_counts()

1957                          11
1942                           9
1956                           8
1958                           7
1950                           7
                              ..
01-Dec-1994                    1
Reported      10-Dec-1994      1
11-Dec-1994                    1
13-Dec-1994                    1
1845-1853                      1
Name: Date, Length: 5433, dtype: int64

In [26]:
mask = data.isnull()            # Me devuelve una mascara booleana de toda la tabla.

In [27]:
null_values_count = mask.T.sum()  # Me devuelve el nº de NaN que hay por fila.

In [28]:
null_values_count

0         2
1         3
2         3
3         4
4         4
         ..
6306     22
6307     22
6308     22
6309     23
25722    23
Length: 6311, dtype: int64

In [29]:
nan_por_linea_100 = (null_values_count/len(data.T)) *100

In [30]:
nan_por_linea_100.min()

8.333333333333332

In [31]:
mask2 = nan_por_linea_100 > 50 

In [32]:
mask2

0        False
1        False
2        False
3        False
4        False
         ...  
6306      True
6307      True
6308      True
6309      True
25722     True
Length: 6311, dtype: bool

In [33]:
data = data[~mask2]

In [34]:
data.shape

(6302, 24)

In [35]:
data.isnull().T.sum()

0       2
1       3
2       3
3       4
4       4
       ..
6297    5
6298    6
6299    5
6300    7
6301    4
Length: 6302, dtype: int64

In [36]:
'''Utilizo la función to_datetime para convertir los valores de la columna 'Date' a objetos de fecha, y utilizo
el parámetro "errors='coerce'" para que los valores que no se pueden convertir se reemplacen con NaT (Not a Time).
'''
#data['Date'] = pd.to_datetime(data['Date'], errors='coerce')

'Utilizo la función to_datetime para convertir los valores de la columna \'Date\' a objetos de fecha, y utilizo\nel parámetro "errors=\'coerce\'" para que los valores que no se pueden convertir se reemplacen con NaT (Not a Time).\n'

In [37]:
print(data['Date'].dtypes)

object


In [38]:
data['Year'].is_monotonic_decreasing

False

In [39]:
data.head()

Case_Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex  Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                                      Injury Fatal  \
0  No injury to occupant, outrigger canoe and paddle damaged     N   
1                                 Minor injury to left thigh     N   
2               Injury to left lower leg from surfboard skeg     N   
3                                  Minor injury to lower leg     N   
4          Lacerations to leg & hand shark PROVOKED INCIDENT     N   

            Time          Species          Investigator_or_Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                                                     href_formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1    http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.18-McNeely.pdf   
2     http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.09-Denges.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.08-Arrawarra.pdf   
4      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.04-Ramos.pdf   

  Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
0    2018.06.25    2018.06.25          6303.0        NaN        NaN  
1    2018.06.18    2018.06.18          6302.0        NaN        NaN  
2    2018.06.09    2018.06.09          6301.0        NaN        NaN  
3    2018.06.08    2018.06.08          6300.0        NaN        NaN  
4    2018.06.04    2018.06.04          6299.0        NaN        NaN

In [40]:
data['Date'].iloc[-150:-100]

6152                        1749
6153                        1755
6154                        1748
6155                 17-Dec-1742
6156        Reported 06-Apr-1738
6157                        1733
6158                        1723
6159                   June 1721
6160                 26-Mar-1703
6161                       1700s
6162                       1700s
6163                       1700s
6164    Late 1600s Reported 1728
6165               Reported 1638
6166               Reported 1637
6167               Reported 1617
6168                        1642
6169                        1595
6170    Letter dated 10-Jan-1580
6171                        1555
6172                    Ca. 1554
6173                    Ca. 1543
6174              Circa 500 A.D.
6175                    77  A.D.
6176                  Ca. 5 A.D.
6177                Ca. 214 B.C.
6178               Ca. 336.B.C..
6179                    493 B.C.
6180                Ca. 725 B.C.
6181                 Before 1939
6182      

In [41]:
data = data.sort_values(by='Year',ascending=False)
data = data.reset_index(drop=True)


In [42]:
data.iloc[-150:-100]

Case_Number                                    Date    Year  \
6152    1748.00.00                                    1748  1748.0   
6153    1742.12.17                             17-Dec-1742  1742.0   
6154  1738.04.06.R                    Reported 06-Apr-1738  1738.0   
6155    1733.00.00                                    1733  1733.0   
6156    1723.00.00                                    1723  1723.0   
6157    1721.06.00                               June 1721  1721.0   
6158    1703.03.26                             26-Mar-1703  1703.0   
6159  1700.00.00.c                                   1700s  1700.0   
6160  1700.00.00.b                                   1700s  1700.0   
6161  1700.00.00.a                                   1700s  1700.0   
6162  1642.00.00.b                Late 1600s Reported 1728  1642.0   
6163    1642.00.00                                    1642  1642.0   
6164  1638.00.00.R                           Reported 1638  1638.0   
6165  1637.00.00.R                           Reported 1637  1637.0   
6166  1617.00.00.R                           Reported 1617  1617.0   
6167    1595.00.00                                    1595  1595.0   
6168  1580.01.10.R                Letter dated 10-Jan-1580  1580.0   
6169    1555.00.00                                    1555  1555.0   
6170    1554.00.00                                Ca. 1554  1554.0   
6171    1543.00.00                                Ca. 1543  1543.0   
6172    0500.00.00                          Circa 500 A.D.   500.0   
6173    0077.00.00                                77  A.D.    77.0   
6174    0005.00.00                              Ca. 5 A.D.     5.0   
6175       ND.0041  Said to be 1941-1945, more likely 1945     0.0   
6176       ND.0047                         Before Mar-1956     0.0   
6177       ND.0046                             Before 1952     0.0   
6178       ND.0044                               1941-1945     0.0   
6179       ND.0043              "During the war" 1943-1945     0.0   
6180       ND.0042                        "Before the war"     0.0   
6181       ND.0038                               1941-1942     0.0   
6182       ND.0040                               1941-1945     0.0   
6183       ND.0039                               1941-1945     0.0   
6184       ND.0049                             Before 1956     0.0   
6185       ND.0037                             1940 - 1950     0.0   
6186       ND.0036                             1940 - 1950     0.0   
6187       ND.0035                             1940 - 1950     0.0   
6188       ND.0048                             Before 1956     0.0   
6189       ND.0052                             Before 1957     0.0   
6190       ND.0051                             Before 1957     0.0   
6191       ND.0033                             Before 1905     0.0   
6192       ND.0053                             Before 1958     0.0   
6193       ND.0054                             Before 1958     0.0   
6194       ND.0055                             Before 1958     0.0   
6195       ND.0056                             Before 1958     0.0   
6196       ND.0057                            Before  1958     0.0   
6197       ND.0058                            Before  1958     0.0   
6198       ND.0059                     Before  24 Apr-1959     0.0   
6199       ND.0060                     Before  19-Jun-1959     0.0   
6200       ND.0062                             Before 1960     0.0   
6201       ND.0063                                   1960s     0.0   

              Type                   Country                          Area  \
6152    Unprovoked                    PANAMA        Las Perlas archipelago   
6153    Unprovoked                       NaN                           NaN   
6154    Unprovoked                     ITALY                        Sicily   
6155       Invalid                   ICELAND                   Bardestrand   
6156    Unprovoked                    ROATAN  

In [43]:
data['Year'].loc[:6171].is_monotonic_decreasing

True

In [44]:
data = data.loc[:6171]

In [45]:
print(data['Date'].iloc[-100:-80])

6072              14-Apr-1839
6073                1839/1840
6074                 Ca. 1839
6075              17-Jan-1837
6076                 Ca. 1837
6077                 Jul-1837
6078              09-Sep-1837
6079                 1836.00.
6080             1836.07.26.R
6081     Reported 21-Feb-1835
6082              06-Jan-1835
6083     Reported 15-Jul-1834
6084              04-Jun-1832
6085     Reported 23-Jan-1832
6086              24-Aug-1831
6087    Reported 22- Jan-1831
6088     Reported 22-Apr-1830
6089              30-Apr-1830
6090     Reported 02-Jul-1830
6091              26-Jul-1830
Name: Date, dtype: object


In [46]:

data['Date'] = pd.to_datetime(data['Date'], errors='coerce').dt.month


In [47]:
print(data['Date'].iloc[-100:-80])

6072    4.0
6073    NaN
6074    NaN
6075    1.0
6076    NaN
6077    7.0
6078    9.0
6079    NaN
6080    NaN
6081    NaN
6082    1.0
6083    NaN
6084    6.0
6085    NaN
6086    8.0
6087    NaN
6088    NaN
6089    4.0
6090    NaN
6091    7.0
Name: Date, dtype: float64


In [48]:
data['Date'].isna().value_counts()

False    5458
True      714
Name: Date, dtype: int64

In [49]:
data['Date'].fillna(data['Date'].mode()[0], inplace=True)

In [50]:
print(data['Date'].iloc[-100:-90])

6072    4.0
6073    1.0
6074    1.0
6075    1.0
6076    1.0
6077    7.0
6078    9.0
6079    1.0
6080    1.0
6081    1.0
Name: Date, dtype: float64


In [51]:
data.rename(columns={'Date': 'Month'}, inplace=True)

In [52]:
data['Month'].loc[4537]

1.0

In [53]:
data['Month'] = data['Month'].astype(int)

In [54]:
data['Case_Number'].iloc[1490:1545]

1490      2006.09.02
1491    2006.07.31.R
1492    2006.08.29.b
1493      2006.07.12
1494      2006.07.13
1495      2006.07.17
1496    2006.07.17.R
1497      2006.07.23
1498      2006.07.25
1499      2006.07.28
1500    2006.07.29.a
1501    2006.07.29.b
1502    2006.07.31.a
1503      2006.07.10
1504    2006.08.00.a
1505    2006.08.00.b
1506      2006.08.13
1507      2006.08.15
1508    2006.08.20.a
1509    2006.08.20.b
1510    2006.08.22.a
1511    2006.08.22.b
1512      2006.08.27
1513    2006.08.29.a
1514      2005.05.25
1515      2005.06.04
1516      2005.06.02
1517      2005.05.28
1518      2005.05.27
1519      2005.05.02
1520      2005.05.15
1521      2005.05.14
1522      2005.05.03
1523    2005.05.02.R
1524      2005.06.07
1525      2005.06.05
1526      2005.06.22
1527      2005.06.13
1528      2005.06.16
1529      2005.06.18
1530      2005.06.21
1531    2005.04.17.R
1532      2005.06.25
1533      2005.07.01
1534      2005.07.13
1535      2005.07.15
1536    2005.07.15.R
1537    2005.

In [55]:
data['Case_Number'] = data['Case_Number'].str.replace(r'[a-zA-Z]', '')

In [56]:
data['Case_Number'] = pd.to_datetime(data['Case_Number'], errors='coerce').dt.day

In [57]:
import random

days = []

for i in data.index:
    
    if pd.isnull(data['Case_Number'].iloc[i]):
        
        days.append(random.randint(1,28))
    else:
        
        days.append(data['Case_Number'].iloc[i])
        
data['Case_Number'] = days

In [58]:
data.at[5403, 'Case_Number']

24.0

In [59]:
data['Case_Number'].iloc[5403]

24.0

In [60]:
data.rename(columns={'Case_Number': 'Day'}, inplace=True)

In [61]:
data['Day'] = data['Day'].astype(int)

In [62]:
data.original_order.iloc[-10:]

6162    139.0
6163    135.0
6164    138.0
6165    137.0
6166    136.0
6167    134.0
6168    133.0
6169    132.0
6170    131.0
6171    130.0
Name: original_order, dtype: float64

In [63]:
data['original_order'] = data['original_order'].astype(int)

In [64]:
data['Year'] = data['Year'].astype(int)

In [65]:
data.head()

Day  Month  Year        Type    Country               Area  \
0   25      6  2018     Boating        USA         California   
1   23      2  2018  Unprovoked  AUSTRALIA    New South Wales   
2   15      4  2018  Unprovoked  AUSTRALIA  Western Australia   
3   14      4  2018  Unprovoked    BAHAMAS     New Providence   
4   10      1  2018     Invalid     BRAZIL            Alagoas   

                             Location  Activity                  Name Sex  \
0         Oceanside, San Diego County  Paddling           Julie Wolfe   F   
1  Little Congwong Beach, La Perouse   Swimming        Anna Shurapey    F   
2   Cobblestones, Margaret River Area   Surfing  Alejandro Travaglini   M   
3                       Nirvana Beach   Surfing           Bruce Rowan   M   
4           Praia de Sauaçuhy, Maceió   Fishing            Josias Paz   M   

   Age  \
0   57   
1   55   
2   37   
3  NaN   
4   56   

                                                                   Injury  \
0               No injury to occupant, outrigger canoe and paddle damaged   
1                                          Laceratons to right leg & foot   
2                                                     Lacerations to legs   
3                          No Injury. Shark swam away with the surf board   
4  Injury to ankle from marine animal trapped in weir PROVOKED INCIDENT.    

  Fatal         Time                              Species  \
0     N        18h00                          White shark   
1     N  19h00, Dusk  Juvenile white shark,  2.7 to 3.2 m   
2     N        08h00                                  NaN   
3     N        09h30                          Tiger shark   
4     N          NaN      Shark involvement not confirmed   

            Investigator_or_Source                          pdf  \
0                 R. Collier, GSAF         2018.06.25-Wolfe.pdf   
1                   B. Myatt, GSAF      2018.02.23-Shurapey.pdf   
2                    B.Myatt, GSAF  2018.04.15.a-Travaglini.pdf   
3                      Tribune242,         2018.04.14-Rowan.pff   
4  K. McMurray, TrackingSharks.com         2018.04.10.R-Paz.pdf   

                                                                        href_formula  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

                                                                                href  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

  Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
0    2018.06.25    2018.06.25            6303        NaN        NaN  
1    2018.02.23    2018.02.23            6263        NaN        NaN  
2  2018.04.15.a  2018.04.15.a            6274        NaN        NaN  
3    2018.04.14    2018.04.14            6273        NaN        NaN  
4  2018.04.10.R  2018.04.10.R            6272        NaN        NaN

In [66]:
data.Type.value_counts()

Unprovoked      4483
Provoked         567
Invalid          544
Sea Disaster     233
Boating          203
Boat             135
Questionable       2
Boatomg            1
Name: Type, dtype: int64

In [67]:
data['Type'] = data['Type'].replace({'Boatomg': 'Boat'})

In [68]:
data['Type'] = data['Type'].replace({'Boat': 'Boating'})

In [69]:
data.Type.value_counts()

Unprovoked      4483
Provoked         567
Invalid          544
Boating          339
Sea Disaster     233
Questionable       2
Name: Type, dtype: int64

In [70]:
data.Type.dtype

dtype('O')

In [71]:
data[data['Type'] == 'Boatomg']

Empty DataFrame
Columns: [Day, Month, Year, Type, Country, Area, Location, Activity, Name, Sex, Age, Injury, Fatal, Time, Species, Investigator_or_Source, pdf, href_formula, href, Case_Number_1, Case_Number_2, original_order, Unnamed_22, Unnamed_23]
Index: []

In [72]:
data[data['Type'] == 'Questionable']

Day  Month  Year          Type    Country             Area  \
47   25      4  2018  Questionable  AUSTRALIA  New South Wales   
51    9      5  2018  Questionable  AUSTRALIA  New South Wales   

                  Location Activity         Name Sex  Age  \
47             Lennox Head  Surfing  Matthew Lee   M  NaN   
51  Sharpes Beach, Ballina  Surfing         male   M  NaN   

                          Injury Fatal   Time  \
47                     No injury     N  07h00   
51  No injury, surfboard damaged     N  10h30   

                            Species Investigator_or_Source  \
47                     Questionable         B. Myatt, GSAF   
51  Shark involvement not confirmed         B. Myatt, GSAF   

                            pdf  \
47         2018.04.25.b-Lee.pdf   
51  2018.05.09-SharpesBeach.pdf   

                                                                         href_formula  \
47         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.25.b-Lee.pdf   
51  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.05.09-SharpesBeach.pdf   

                                                                                 href  \
47         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.25.b-Lee.pdf   
51  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.05.09-SharpesBeach.pdf   

   Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
47  2018.04.25.b  2018.04.25.b            6283        NaN        NaN  
51    2018.05.09    2018.05.09            6287        NaN        NaN

In [73]:
data = data.drop(data[data['Type'] == 'Questionable'].index)

In [74]:
data = data.reset_index(drop=True)

In [75]:
data.Country.unique()

array(['USA', 'AUSTRALIA', 'BAHAMAS', 'BRAZIL', 'NEW CALEDONIA',
       'SOUTH AFRICA', 'ECUADOR', 'THAILAND', 'ENGLAND', 'MEXICO',
       'MALDIVES', 'COSTA RICA', 'UNITED ARAB EMIRATES',
       'ST HELENA, British overseas territory', 'REUNION', 'NEW ZEALAND',
       'UNITED KINGDOM', 'FRENCH POLYNESIA', 'SPAIN', 'COMOROS',
       'INDONESIA', 'PHILIPPINES', 'MAURITIUS', 'LIBYA', nan, 'CUBA',
       'SAMOA', 'MALAYSIA', 'EGYPT', 'SOLOMON ISLANDS', 'JAPAN',
       'COLUMBIA', 'CAPE VERDE', 'CAYMAN ISLANDS', 'DOMINICAN REPUBLIC',
       'Fiji', 'CHINA', 'PUERTO RICO', 'ATLANTIC OCEAN', 'ITALY',
       'MOZAMBIQUE', 'ARUBA', 'FIJI', 'FRANCE', 'ST. MARTIN',
       'TRINIDAD & TOBAGO', 'PAPUA NEW GUINEA', 'GREECE',
       'PALESTINIAN TERRITORIES', 'JAMAICA', 'TAIWAN', 'DIEGO GARCIA',
       'BELIZE', 'SEYCHELLES', 'GUAM', 'ISRAEL', 'KIRIBATI', 'CHILE',
       'SAUDI ARABIA', 'CROATIA', 'NIGERIA', 'TONGA', 'CANADA',
       'SCOTLAND', 'TURKS & CAICOS', 'UNITED ARAB EMIRATES (UAE)',
      

In [76]:
data['Country'] = data['Country'].replace({'Fiji': 'FIJI', 'ST HELENA, British overseas territory': 'UNITED KINGDOM OVERSEAS', 'DIEGO GARCIA': 'UNITED KINGDOM OVERSEAS', 'Sierra Leone': 'SIERRA LEONE', 'Seychelles': 'SEYCHELLES',
                                 'EGYPT / ISRAEL': 'EGYPT', 'PACIFIC OCEAN ': 'PACIFIC OCEAN', 'BRITISH ISLES': 'UNITED KINGDOM', 'ENGLAND': 'UNITED KINGDOM', 'ST. MAARTIN': 'ST MARTIN', 'ST. MARTIN': 'ST MARTIN',
                                'NORTH ATLANTIC OCEAN ': 'NORTH ATLANTIC OCEAN', 'FEDERATED STATES OF MICRONESIA': 'MICRONESIA', 'BRITISH WEST INDIES': 'UNITED KINGDOM OVERSEAS', 'RED SEA / INDIAN OCEAN': 'RED SEA', 'ANDAMAN / NICOBAR ISLANDAS': 'BAY OF BENGAL',
                                 'SUDAN?': 'SUDAN', 'THE BALKANS': 'SLOVENIA', 'IRAN / IRAQ': 'IRAN', ' PHILIPPINES': 'PHILIPPINES', 'SOLOMON ISLANDS / VANUATU': 'VANUATU', 'ITALY / CROATIA': 'CROATIA', 'YEMEN ': 'YEMEN', 'REUNION': 'REUNION ISLAND',
                                'EGYPT ': 'EGYPT', 'BRITISH NEW GUINEA': 'UNITED KINGDOM OVERSEAS', 'OCEAN': 'PACIFIC OCEAN', 'INDIAN OCEAN?': 'INDIAN OCEAN', 'EQUATORIAL GUINEA / CAMEROON': 'CAMEROON', 'Coast of AFRICA': 'ATLANTIC OCEAN', 'Between PORTUGAL & INDIA': 'INDIAN OCEAN',
                                'TURKS & CAICOS': 'UNITED KINGDOM OVERSEAS', 'TRINIDAD & TOBAGO': 'TOBAGO', 'UNITED ARAB EMIRATES (UAE)': 'UNITED ARAB EMIRATES', 'BRITISH VIRGIN ISLANDS': 'UNITED KINGDOM OVERSEAS', ' TONGA': 'TONGA', 'MEXICO ': 'MEXICO', 'NICARAGUA ': 'NICARAGUA',
                                 'MID-PACIFC OCEAN': 'MID PACIFIC OCEAN'})
                                  

In [77]:
data['Country'].fillna(value='unknown', inplace=True) 

In [78]:
data.Country.isna().unique()

array([False])

In [79]:
data['Type'].fillna(value='Invalid', inplace=True)

In [80]:
data = data.drop(6047)

data = data.reset_index(drop=True)

In [81]:
data.head()

Day  Month  Year        Type    Country               Area  \
0   25      6  2018     Boating        USA         California   
1   23      2  2018  Unprovoked  AUSTRALIA    New South Wales   
2   15      4  2018  Unprovoked  AUSTRALIA  Western Australia   
3   14      4  2018  Unprovoked    BAHAMAS     New Providence   
4   10      1  2018     Invalid     BRAZIL            Alagoas   

                             Location  Activity                  Name Sex  \
0         Oceanside, San Diego County  Paddling           Julie Wolfe   F   
1  Little Congwong Beach, La Perouse   Swimming        Anna Shurapey    F   
2   Cobblestones, Margaret River Area   Surfing  Alejandro Travaglini   M   
3                       Nirvana Beach   Surfing           Bruce Rowan   M   
4           Praia de Sauaçuhy, Maceió   Fishing            Josias Paz   M   

   Age  \
0   57   
1   55   
2   37   
3  NaN   
4   56   

                                                                   Injury  \
0               No injury to occupant, outrigger canoe and paddle damaged   
1                                          Laceratons to right leg & foot   
2                                                     Lacerations to legs   
3                          No Injury. Shark swam away with the surf board   
4  Injury to ankle from marine animal trapped in weir PROVOKED INCIDENT.    

  Fatal         Time                              Species  \
0     N        18h00                          White shark   
1     N  19h00, Dusk  Juvenile white shark,  2.7 to 3.2 m   
2     N        08h00                                  NaN   
3     N        09h30                          Tiger shark   
4     N          NaN      Shark involvement not confirmed   

            Investigator_or_Source                          pdf  \
0                 R. Collier, GSAF         2018.06.25-Wolfe.pdf   
1                   B. Myatt, GSAF      2018.02.23-Shurapey.pdf   
2                    B.Myatt, GSAF  2018.04.15.a-Travaglini.pdf   
3                      Tribune242,         2018.04.14-Rowan.pff   
4  K. McMurray, TrackingSharks.com         2018.04.10.R-Paz.pdf   

                                                                        href_formula  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

                                                                                href  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

  Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
0    2018.06.25    2018.06.25            6303        NaN        NaN  
1    2018.02.23    2018.02.23            6263        NaN        NaN  
2  2018.04.15.a  2018.04.15.a            6274        NaN        NaN  
3    2018.04.14    2018.04.14            6273        NaN        NaN  
4  2018.04.10.R  2018.04.10.R            6272        NaN        NaN

In [82]:
data.Location.unique()

array(['Oceanside, San Diego County',
       'Little Congwong Beach, La Perouse ',
       'Cobblestones, Margaret River Area', ..., 'Hooghly River mouth',
       'Ganges Delta', 'River Cochin'], dtype=object)

In [83]:
data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6169 entries, 0 to 6168
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Day                     6169 non-null   int32 
 1   Month                   6169 non-null   int32 
 2   Year                    6169 non-null   int32 
 3   Type                    6169 non-null   object
 4   Country                 6169 non-null   object
 5   Area                    5742 non-null   object
 6   Location                5667 non-null   object
 7   Activity                5645 non-null   object
 8   Name                    5964 non-null   object
 9   Sex                     5612 non-null   object
 10  Age                     3457 non-null   object
 11  Injury                  6142 non-null   object
 12  Fatal                   5632 non-null   object
 13  Time                    2939 non-null   object
 14  Species                 3426 non-null   object
 15  Inve

In [84]:
len(data[(data.Area.isna() == True) | (data.Location.isna() == True)])

712

In [85]:
len(data[(data.Area.isna() == True) & (data.Location.isna() == True)])

217

In [86]:
doble_nan = (data.Area.isna() == True) & (data.Location.isna() == True)

data.loc[doble_nan, ["Area","Location"]] = data.loc[doble_nan, ["Area","Location"]].fillna("unknown")

In [87]:
data.head()

Day  Month  Year        Type    Country               Area  \
0   25      6  2018     Boating        USA         California   
1   23      2  2018  Unprovoked  AUSTRALIA    New South Wales   
2   15      4  2018  Unprovoked  AUSTRALIA  Western Australia   
3   14      4  2018  Unprovoked    BAHAMAS     New Providence   
4   10      1  2018     Invalid     BRAZIL            Alagoas   

                             Location  Activity                  Name Sex  \
0         Oceanside, San Diego County  Paddling           Julie Wolfe   F   
1  Little Congwong Beach, La Perouse   Swimming        Anna Shurapey    F   
2   Cobblestones, Margaret River Area   Surfing  Alejandro Travaglini   M   
3                       Nirvana Beach   Surfing           Bruce Rowan   M   
4           Praia de Sauaçuhy, Maceió   Fishing            Josias Paz   M   

   Age  \
0   57   
1   55   
2   37   
3  NaN   
4   56   

                                                                   Injury  \
0               No injury to occupant, outrigger canoe and paddle damaged   
1                                          Laceratons to right leg & foot   
2                                                     Lacerations to legs   
3                          No Injury. Shark swam away with the surf board   
4  Injury to ankle from marine animal trapped in weir PROVOKED INCIDENT.    

  Fatal         Time                              Species  \
0     N        18h00                          White shark   
1     N  19h00, Dusk  Juvenile white shark,  2.7 to 3.2 m   
2     N        08h00                                  NaN   
3     N        09h30                          Tiger shark   
4     N          NaN      Shark involvement not confirmed   

            Investigator_or_Source                          pdf  \
0                 R. Collier, GSAF         2018.06.25-Wolfe.pdf   
1                   B. Myatt, GSAF      2018.02.23-Shurapey.pdf   
2                    B.Myatt, GSAF  2018.04.15.a-Travaglini.pdf   
3                      Tribune242,         2018.04.14-Rowan.pff   
4  K. McMurray, TrackingSharks.com         2018.04.10.R-Paz.pdf   

                                                                        href_formula  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

                                                                                href  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

  Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
0    2018.06.25    2018.06.25            6303        NaN        NaN  
1    2018.02.23    2018.02.23            6263        NaN        NaN  
2  2018.04.15.a  2018.04.15.a            6274        NaN        NaN  
3    2018.04.14    2018.04.14            6273        NaN        NaN  
4  2018.04.10.R  2018.04.10.R            6272        NaN        NaN

In [88]:
data.Area.fillna(data.Location, inplace=True)

In [89]:
data.Location.fillna(data.Area, inplace=True)

In [90]:
data.Activity.value_counts().head(60)

Surfing                           968
Swimming                          853
Fishing                           423
Spearfishing                      332
Bathing                           159
Wading                            147
Diving                            115
Standing                           97
Snorkeling                         88
Scuba diving                       75
Body boarding                      61
Body surfing                       49
Swimming                           47
Kayaking                           33
Fell overboard                     32
Treading water                     32
Boogie boarding                    29
Pearl diving                       28
Free diving                        27
Windsurfing                        19
Walking                            17
Boogie Boarding                    16
Shark fishing                      15
Floating                           14
Fishing                            13
Rowing                             12
Surf fishing

In [91]:
data['Activity'] = np.where(data['Activity'].str.contains('urf'), 'Surfing', data['Activity'])

In [92]:
data['Activity'] = np.where(data['Activity'].str.contains('ishin'), 'Fishing', data['Activity'])

In [93]:
data['Activity'] = np.where(data['Activity'].str.contains('iving'), 'Diving', data['Activity'])

In [94]:
data['Activity'] = np.where(data['Activity'].str.contains('wim'), 'Swimming', data['Activity'])

In [95]:
data['Activity'] = np.where(data['Activity'].str.contains('ath'), 'Bathing', data['Activity'])

In [96]:
data['Activity'] = np.where(data['Activity'].str.contains('oard'), 'Boarding', data['Activity'])

In [97]:
data['Activity'] = np.where(data['Activity'].str.contains('ayak'), 'Boating', data['Activity'])

In [98]:
data['Activity'] = np.where(data['Activity'].str.contains('Boat'), 'Boating', data['Activity'])

In [99]:
data['Activity'] = np.where(data['Activity'].str.contains('boat'), 'Boating', data['Activity'])

In [100]:
data['Activity'] = np.where(data['Activity'].str.contains('Walking'), 'Wading', data['Activity'])

In [101]:
data['Activity'] = np.where(data['Activity'].str.contains('Standing'), 'Standing', data['Activity'])

In [102]:
data['Activity'] = np.where(data['Activity'].str.contains('loating'), 'Swimming', data['Activity'])

In [103]:
data['Activity'] = np.where(data['Activity'].str.contains('Canoe'), 'Boating', data['Activity'])

In [104]:
data['Activity'] = np.where(data['Activity'].str.contains('Sail'), 'Boating', data['Activity'])

In [105]:
data['Activity'].fillna(value='unknown', inplace=True)

In [106]:
excepts = {'Surfing':'Surfing', 'Fishing':'Fishing', 'Swimming':'Swimming', 'Diving':'Diving', 'Boarding':'Boarding', 
                'Bathing':'Bathing', 'Wading':'Wading', 'Boating':'Boating', 'Standing':'Standing', 'Snorkeling':'Snorkeling'}

data.loc[~data['Activity'].isin(excepts), 'Activity'] = 'unknown'


In [107]:
data['Name'].fillna(value='unknown', inplace=True)

In [108]:
data.Sex.value_counts()

M      4982
F       624
M         2
N         2
lli       1
.         1
Name: Sex, dtype: int64

In [109]:
data[data['Sex'] == 'N']

Day  Month  Year      Type    Country             Area  Location  \
4933   11      7  1934   Boating  AUSTRALIA  New South Wales  Cronulla   
6126   18      1  1801  Provoked    unknown          unknown   unknown   

      Activity                                         Name Sex  Age  \
4933   Fishing  18' boat, occupants William & Leslie Newton   N  NaN   
6126  Standing                             Stephen Pettigew   N  NaN   

                                                                                          Injury  \
4933  No injury to occupants Sharks continually followed the dinghy, and one smashed its rudder    
6126                                                                    FATAL, PROVOKED INCIDENT   

     Fatal Time             Species  \
4933     N  NaN  Blue pointer, 11'    
6126     Y  NaN           12' shark   

                                                                Investigator_or_Source  \
4933  G.P. Whitley, ref: Daily Telegraph, 7/11/1934 & Sydney Morning Herald  7/12/1934   
6126                                                      The Evening Post, 12/18/1801   

                                       pdf  \
4933  1934.07.11-Newton-boat-Australia.pdf   
6126            1801.12.18.R-Pettigrew.pdf   

                                                                                    href_formula  \
4933  http://sharkattackfile.net/spreadsheets/pdf_directory/1934.07.11-Newton-boat-Australia.pdf   
6126            http://sharkattackfile.net/spreadsheets/pdf_directory/1801.12.18.R-Pettigrew.pdf   

                                                                                            href  \
4933  http://sharkattackfile.net/spreadsheets/pdf_directory/1934.07.11-Newton-boat-Australia.pdf   
6126            http://sharkattackfile.net/spreadsheets/pdf_directory/1801.12.18.R-Pettigrew.pdf   

     Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
4933    1934.07.11    1934.07.11            1365        NaN        NaN  
6126  1801.12.18.R  1801.12.18.R             172        NaN        NaN

In [110]:
data['Sex'] = data['Sex'].replace({'M ': 'M', 'lli': 'M', 'N': 'M', '.': 'unknown'})
data['Sex'].fillna(value='unknown', inplace=True)

In [111]:
data.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6169 entries, 0 to 6168
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Day                     6169 non-null   int32 
 1   Month                   6169 non-null   int32 
 2   Year                    6169 non-null   int32 
 3   Type                    6169 non-null   object
 4   Country                 6169 non-null   object
 5   Area                    6169 non-null   object
 6   Location                6169 non-null   object
 7   Activity                6169 non-null   object
 8   Name                    6169 non-null   object
 9   Sex                     6169 non-null   object
 10  Age                     3457 non-null   object
 11  Injury                  6142 non-null   object
 12  Fatal                   5632 non-null   object
 13  Time                    2939 non-null   object
 14  Species                 3426 non-null   object
 15  Inve

In [112]:
len(data[data.Age.isna()==True])

2712

In [113]:
data.Age.value_counts().head(60)

17      154
18      150
19      141
20      140
15      137
16      136
21      118
22      117
25      107
24      106
14      101
13       94
26       83
28       80
23       80
29       78
27       77
30       76
12       73
32       69
35       68
40       56
10       56
31       52
34       50
38       48
33       44
43       43
37       42
36       42
39       38
42       38
41       38
11       37
52       35
9        35
50       32
45       32
47       30
44       29
48       28
49       28
8        28
46       27
55       23
7        22
51       20
54       17
60       16
57       16
58       15
53       13
61       13
6        13
59       11
56       10
69       10
63        9
Teen      9
62        8
Name: Age, dtype: int64

In [114]:
data['Age'] = data['Age'].str.replace(' ', '')

In [115]:
data['Age'] = data['Age'].str.replace('[^0-9]', '')

In [116]:
data['Age'] = data['Age'].astype(str)   # Convierto los valores de 'Age' a str para poder aplicar la función

def no_age_unknown(x):
    
    if len(x) > 2:
        
        return 'unknown'
    
    return x

data['Age'] = data['Age'].apply(no_age_unknown)

In [117]:
data['Age'].replace("", "unknown", inplace=True)

In [118]:
data.info(memory_usage = 'deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6169 entries, 0 to 6168
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Day                     6169 non-null   int32 
 1   Month                   6169 non-null   int32 
 2   Year                    6169 non-null   int32 
 3   Type                    6169 non-null   object
 4   Country                 6169 non-null   object
 5   Area                    6169 non-null   object
 6   Location                6169 non-null   object
 7   Activity                6169 non-null   object
 8   Name                    6169 non-null   object
 9   Sex                     6169 non-null   object
 10  Age                     6169 non-null   object
 11  Injury                  6142 non-null   object
 12  Fatal                   5632 non-null   object
 13  Time                    2939 non-null   object
 14  Species                 3426 non-null   object
 15  Inve

In [119]:
data.Injury.value_counts().tail(60)

Left inner thigh                                                                                                          1
Sharks prevented recovery of remains                                                                                      1
Minor laceration & 3 punctures to right foot                                                                              1
No injury, flung off board                                                                                                1
No injury to occupant; shark bit propeller                                                                                1
No injury, wetsuit punctured                                                                                              1
Fingers bitten PROVOKED INCIDENT                                                                                          1
Puncture wounds on knee                                                                                                   1
Minor in

In [120]:
data['Injury'] = np.where(data['Injury'].str.contains('FATAL'), 'FATAL', data['Injury'])

In [121]:
data['Injury'] = np.where(data['Injury'].str.contains('atal'), 'FATAL', data['Injury'])

In [122]:
data['Injury'].fillna(value='unknown', inplace=True)

In [123]:
data[data.Fatal.isna()==True]

Day  Month  Year        Type  Country            Area  \
14     15      4  2018  Unprovoked   BRAZIL      Pernambuco   
56      3      5  2017     Invalid      USA      California   
62     26      4  2017     Invalid      USA         Florida   
64     20      4  2017     Invalid      USA  South Carolina   
79     29      6  2017     Invalid      USA  South Carolina   
...   ...    ...   ...         ...      ...             ...   
6121   16      9  1805     Invalid      USA        New York   
6142   21      1  1767     Invalid   FRANCE    Côte d'Azur    
6152   19      1  1733     Invalid  ICELAND     Bardestrand   
6153    3      1  1723  Unprovoked   ROATAN         unknown   
6159   18      1  1642     Invalid   GUINEA         unknown   

                         Location  Activity                           Name  \
14                        Piedade  Swimming                  Pablo de Melo   
56    Sunset Beach, Orange County   Surfing                    Sophia Raab   
62                        Florida   unknown                  Molly Cavelli   
64              Georgetown County  Swimming                           male   
79             Wrightsville Beach   Fishing                           male   
...                           ...       ...                            ...   
6121   Sag Harbor, Suffolk County   Surfing                        unknown   
6142                   St. Tropez   Bathing                Samuel Matthews   
6152                 Talkknefiord   Surfing                        unknown   
6153                      unknown   Surfing                  Philip Ashton   
6159                      unknown  Boarding  crew member of the Nieuwstadt   

          Sex      Age  \
14          M       34   
56          F       18   
62          F  unknown   
64          M  unknown   
79          M  unknown   
...       ...      ...   
6121        M  unknown   
6142        M  unknown   
6152  unknown  unknown   
6153        M  unknown   
6159        M  unknown   

                                                                               Injury  \
14    Multiple severe injuries to arms and leg, leg subsequently surgically amputated   
56                                Laceration to thigh, likely caused by surfboard fin   
62                                                   Alleged laceration to left ankle   
64                                          Laceration & puncture wounds to left foot   
79                                                Arm injured by hook, not by a shark   
...                                                                               ...   
6121                                        human remains (male) found in shark’s gut   
6142                                                         Lacerations to arm & leg   
6152   Partial hominid remains recovered from shark, probable drowning and scavenging   
6153                                                                  Struck on thigh   
6159                                                                            FATAL   

     Fatal   Time                                         Species  \
14     NaN    NaN                                             NaN   
56     NaN  14h30               Shark involvement highly doubtful   
62     NaN    NaN  No shark invovlement - it ws a publicity stunt   
64     NaN  08h50                 Shark involvement not confirmed   
79     NaN  19h30                           No shark invovlement    
...    ...    ...                                             ...   
6121   NaN    NaN    Shark involvement prior to death unconfirmed   
6142   NaN    NaN         Description of shark does not ring true   
6152   NaN    NaN    Shark involvement prior to death unconfirmed   
6153   NaN    NaN                                             NaN   
6159   NaN    NaN                                    Questionable   

                                 Investigator_or_Source  \
14                                     Glo

In [124]:
data[data['Fatal'] == 'UNKNOWN']

Day  Month  Year        Type    Country               Area  \
701     8      9  2013  Unprovoked        USA     South Carolina   
1203   20      4  2008  Unprovoked  AUSTRALIA    New South Wales   
2208    7      6  1997  Unprovoked     BRAZIL     Rio de Janeiro   
2227   21      2  1997  Unprovoked        USA             Hawaii   
2269   28      4  1996  Unprovoked        USA             Hawaii   
...   ...    ...   ...         ...        ...                ...   
6162   20      1  1637  Unprovoked      INDIA        West Bengal   
6163   22      1  1617  Unprovoked      INDIA        West Bengal   
6164   15      1  1595  Unprovoked      INDIA             Kerala   
6166    6      1  1555  Unprovoked    unknown            unknown   
6167   15      1  1554  Unprovoked     FRANCE  Nice & Marseilles   

                                Location  Activity                   Name  \
701   St. Helena Island, Beaufort County   Surfing                 female   
1203                       Crescent Head   Surfing        Jamie Adlington   
2208          Copacabana, Rio de Janeiro   Bathing      José Luiz Lipiani   
2227                 Sunset Beach, O'ahu   Surfing        Gersome Perreno   
2269                  La'ie Point, O'ahu   Surfing            Wayne Leong   
...                                  ...       ...                    ...   
6162                 Hooghly River mouth    Wading         Hindu pilgrims   
6163                        Ganges Delta   Surfing          Indian people   
6164                        River Cochin   unknown                   male   
6166                             unknown  Swimming                   male   
6167                   Nice & Marseilles   Surfing  males (wearing armor)   

          Sex      Age  \
701         F  unknown   
1203        M  unknown   
2208        M  unknown   
2227        M  unknown   
2269        M  unknown   
...       ...      ...   
6162  unknown  unknown   
6163  unknown  unknown   
6164        M  unknown   
6166        M  unknown   
6167        M  unknown   

                                                                                                   Injury  \
701                                                                                            No details   
1203                                                                                                FATAL   
2208                                                                                                FATAL   
2227                                                                                           No details   
2269                                                                                           No details   
...                                                                                                   ...   
6162                                                                                                FATAL   
6163                                                                                                FATAL   
6164  Leg severed mid-thigh, hand severed, arm above elbow and part of buttocks. Not known if he survived   
6166                                                                                                FATAL   
6167                                                                                                FATAL   

        Fatal Time                Species  \
701   UNKNOWN  NaN                    NaN   
1203  UNKNOWN  NaN     Tiger shark, 2.3m    
2208  UNKNOWN  NaN                    NaN   
2227  UNKNOWN  NaN                    NaN   
2269  UNKNOWN  NaN                    NaN   
...       ...  ...                    ...   
6162  UNKNOWN  NaN                    NaN   
6163  UNKNOWN  NaN                    NaN   
6164  UNKNOWN  NaN                    NaN   
6166  UNKNOWN  NaN                    NaN   
6167  UNKNOWN  NaN  Possibly white sharks   

                           Investigator_or_Source  \
701                              WIS-TV, 9/9/201

In [125]:
data['Fatal'] = data['Fatal'].replace({' N': 'N', 'M': 'N', '2017': 'N', 'N ': 'N', 'y': 'Y'})

In [126]:
data[(data['Injury'] == 'FATAL') & (data['Fatal'] != 'Y')]

Day  Month  Year        Type       Country                   Area  \
116    12      4  2017     Invalid  SOUTH AFRICA          KwaZulu-Natal   
126     1     10  2017     Invalid  SOUTH AFRICA  Western Cape Province   
134    15      9  2017     Invalid  SOUTH AFRICA  Western Cape Province   
372    29      3  2015     Invalid         ITALY               Sardinia   
943    28      8  2011     Invalid     AUSTRALIA             Queensland   
...   ...    ...   ...         ...           ...                    ...   
6161    4      1  1638  Unprovoked       unknown                unknown   
6162   20      1  1637  Unprovoked         INDIA            West Bengal   
6163   22      1  1617  Unprovoked         INDIA            West Bengal   
6166    6      1  1555  Unprovoked       unknown                unknown   
6167   15      1  1554  Unprovoked        FRANCE      Nice & Marseilles   

                 Location  Activity                   Name      Sex      Age  \
116          Protea Banks    Diving      Leopold Mairhuber        M       68   
126           Dyer Island    Diving           Bradley Fick        M       31   
134               Hawston    Diving             Wayon Love        M       25   
372              Sardinia    Diving         Eugenio Masala        M       43   
943        Fantome Island  Swimming                Rooster        M       48   
...                   ...       ...                    ...      ...      ...   
6161              unknown   Surfing                sailors        M  unknown   
6162  Hooghly River mouth    Wading         Hindu pilgrims  unknown  unknown   
6163         Ganges Delta   Surfing          Indian people  unknown  unknown   
6166              unknown  Swimming                   male        M  unknown   
6167    Nice & Marseilles   Surfing  males (wearing armor)        M  unknown   

     Injury    Fatal       Time  \
116   FATAL      NaN        NaN   
126   FATAL      NaN        NaN   
134   FATAL      NaN  Afternoon   
372   FATAL      NaN        NaN   
943   FATAL      NaN      19h30   
...     ...      ...        ...   
6161  FATAL  UNKNOWN        NaN   
6162  FATAL  UNKNOWN        NaN   
6163  FATAL  UNKNOWN        NaN   
6166  FATAL  UNKNOWN        NaN   
6167  FATAL  UNKNOWN        NaN   

                                             Species  \
116   Shark involvement prior to death not confirmed   
126              Death may have been due to drowning   
134                                              NaN   
372                   Shark involvement not cofirmed   
943   Shark involvement prior to death not confirmed   
...                                              ...   
6161                                             NaN   
6162                                             NaN   
6163                                             NaN   
6166                                             NaN   
6167                           Possibly white sharks   

                           Investigator_or_Source  \
116                               E. Ritter, GSAF   
126                       All Africa, 10/11/2017    
134                          Ground Up, 9/20/2017   
372                         A. de Maddalena, GSAF   
943                     Courier Pigeon, 8/30/2011   
...                                           ...   
6161                           Sir Thomas Herbert   
6162  H. Edwards, p.31, citing Sebastian Manrique   
6163      H. Edwards, p.31, citing Samuel Purchas   
6166                                 Olaus Magnus   
6167                                 G. Rondelet    

                                pdf  \
116      2017.04.12.a-Mairhuber.pdf   
126             2017.10.01-Fick.pdf   
134           2017.09.15.b-Love.pdf   
372           2015.03.29-Masala.pdf   
943         2011.08.28-Roosteer.pdf   
...                             ...   
6161           1638.00.00.R-Herbert   
6162      1637.00.00.R-Manrique.pdf   
6163         1617.00.00-Purchas.pdf   
6166  1555.00.00 - Olau

In [127]:
data[(data['Injury'] != 'FATAL') & (data['Fatal'] == 'Y')]

Day  Month  Year          Type                  Country  \
1105    8      2  2009  Sea Disaster                      USA   
1165   13      9  2009      Provoked                   BRAZIL   
1321    4      5  2007  Sea Disaster  UNITED KINGDOM OVERSEAS   
1479   25     11  2006  Sea Disaster              PHILIPPINES   
1621   13      4  2004       Invalid                    TONGA   
...   ...    ...   ...           ...                      ...   
5979   27      3  1860  Sea Disaster             COOK ISLANDS   
5994   21      1  1856    Unprovoked           UNITED KINGDOM   
6096   11      1  1827    Unprovoked                    EGYPT   
6134   26      1  1785    Unprovoked           UNITED KINGDOM   
6136    2      3  1783  Sea Disaster               MARTINIQUE   

                                    Area                             Location  \
1105                         Puerto Rico                         Quebradillas   
1165                          Pernambuco                      Piedade, Recife   
1321                      Providenciales                       Providenciales   
1479                   Surigao del Norte  Off Bilisan Point, Hinatuarn Island   
1621                          Nuku'alofa           30 nautical miles offshore   
...                                  ...                                  ...   
5979                      Mangaia Island                       Mangaia Island   
5994                       Isle of Wight                          Colwell Bay   
6096                          Alexandria                           Alexandria   
6134                              Sussex                             Brighton   
6136  Foundering of the privateer Holker   Foundering of the privateer Holker   

      Activity  \
1105   unknown   
1165   Surfing   
1321   unknown   
1479   unknown   
1621   Fishing   
...        ...   
5979   unknown   
5994  Swimming   
6096   Surfing   
6134   Surfing   
6136   Surfing   

                                                                   Name  \
1105                                           occupant of a Cessna 206   
1165                                         Maurício da Silva Monteiro   
1321  Haitian refugees perished when their boat capsized in choppy seas   
1479                                         Sinking of the m.v.Leonida   
1621                                                             male 1   
...                                                                 ...   
5979                                                  a Cook's Islander   
5994                                                               male   
6096                                                            Two men   
6134                                                            unknown   
6136                                                            unknown   

          Sex      Age  \
1105        M  unknown   
1165        M       34   
1321  unknown  unknown   
1479  unknown  unknown   
1621        M  unknown   
...       ...      ...   
5979        M  unknown   
5994        M  unknown   
6096        M  unknown   
6134        M  unknown   
6136        M  unknown   

                                                                                                   Injury  \
1105        It is probable that all 5 passengers died on impact. The body of one was scavenged by a shark   
1165                                    Cause of death was drowning; his remains were scavenged by sharks   
1321                                               Some of the bodies recovered had been bitten by sharks   
1479                                   15 perished but shark involvement prior to death was not confirmed   
1621  He was was bitten on the arm by small sharks & died, but it was not clear if he died as result o...   
...                                                                                                   ...   
5979                                                      

In [128]:
data['Fatal'] = np.where(data['Injury'] == 'FATAL', 'Y', data['Fatal'])

In [129]:
data.loc[data['Fatal'] == 'Y', 'Injury'] = 'FATAL'

In [130]:
data['Injury'] = np.where(data['Fatal'] == 'UNKNOWN', 'unknown', data['Injury'])

In [131]:
data['Fatal'].replace("UNKNOWN", "unknown", inplace=True)

In [132]:
data.Fatal.value_counts()

N          4230
Y          1410
unknown      51
Name: Fatal, dtype: int64

In [133]:
data.Injury.value_counts().head(20)

FATAL                        1410
Survived                       94
Foot bitten                    83
No injury                      77
Leg bitten                     71
unknown                        52
Left foot bitten               50
Right foot bitten              39
No injury, board bitten        31
Hand bitten                    28
Thigh bitten                   27
Minor injury                   21
Foot lacerated                 21
Lacerations to foot            20
Calf bitten                    20
Right leg bitten               20
Arm bitten                     19
Lacerations to right foot      18
Lacerations to left foot       18
Right calf bitten              16
Name: Injury, dtype: int64

In [134]:
len(data.loc[(data['Injury'] == 'unknown') & (data['Fatal'] == 'unknown')])

51

In [135]:
data.loc[(data['Injury'] == 'unknown') & (data['Fatal'] != 'unknown')]

Day  Month  Year        Type Country       Area         Location  \
4814   21      1  1938  Unprovoked    FIJI  Viti Levu  Singatoka River   

     Activity  Name Sex      Age   Injury Fatal Time Species  \
4814   Wading  male   M  unknown  unknown     N  NaN     NaN   

        Investigator_or_Source                               pdf  \
4814  Time Magazine, 3/21/1938  1938.03.21.R-FijianMethodist.pdf   

                                                                                href_formula  \
4814  http://sharkattackfile.net/spreadsheets/pdf_directory/1938.03.21.R-FijianMethodist.pdf   

                                                                                        href  \
4814  http://sharkattackfile.net/spreadsheets/pdf_directory/1938.03.21.R-FijianMethodist.pdf   

     Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
4814  1938.03.21.R  1938.03.21.R            1483        NaN        NaN

In [136]:
data['Injury'] = np.where(data['Fatal'] == 'N', 'FATAL', data['Injury'])

In [137]:
data['Fatal'].fillna('unknown', inplace=True)

In [138]:
data.Time.value_counts().tail(60)

19h05                                   1
16h12                                   1
17h51                                   1
11h41                                   1
10h45-11h15                             1
15h52                                   1
19h00, Dusk                             1
14h21                                   1
11h25                                   1
10h00 -- 11h00                          1
09h00 -10h00                            1
20h45 (Sunset)                          1
11h56                                   1
08h40                                   1
02h30                                   1
Just after 12h00                        1
Shortly after midnight                  1
Early Morning                           1
11h53                                   1
Shortly before 13h00                    1
12h34                                   1
8:04 pm                                 1
12h46                                   1
Late morning                      

In [139]:
data['Time'] = data['Time'].str.replace(' ', '')

In [140]:
def clean_time(x):
    
    x = x.replace('h', ':').replace('am', '').strip()
    
    if ':' not in x:
        
        if 'unchtime' in x:
            return '12:00'
        elif 'fternoon' in x:
            return '16:00'
        elif 'idnight' in x:
            return '23:59'
        elif 'orning' in x:
            return '09:00'
        elif 'usk' in x:
            return '19:00'
        elif 'vening' in x:
            return '17:00'
        elif 'ight' in x:
            return '21:00'
        elif 'idday' in x:
            return '12:00'
        else:
            return 'unknown'
    else:
        return x

In [141]:
data['Time'] = data['Time'].astype(str)   # Convierto los valores de 'Age' a str para poder aplicar la función

data['Time'] = data['Time'].apply(clean_time)

In [142]:
data['Time'] = data['Time'].str.replace('[^0-9:]', '')  

In [143]:
data['Time'] = data['Time'].replace({':': '', '10:4511:15': '11:00', '07:0008:00': '07:30', '18:1518:30': '18:22', '06:0008:': '07:00',
                                 '17:0017:40': '17:20', ':13:00': '13:00', '14:3015:30': '15:00', '09:0010:00': '09:30', '13:345': '13:34', '9:00': '09:00',
                                '05:0008:00': '06:30', '17:0018:00': '17:30', '10:3013:30': '12:00', '06:0007:00': '06:30', '11:01:': '11:01',
                                 ':03:10': '03:10', '11:0012:00': '11:30', '18:1521:30': '20:00', '10:0014:00': '12:00', '12:0014:00': '13:00', '08:0009:30': '08:45', '09:3015:30': '12:30', '12:4513:45': '13:15',
                                '03:4504:00': '03:52', '15:0015:45': '15:22', '09:3010:00': '09:45', '16:3018:00': '17:15', '8:04': '08:04', '10:0011:00': '10:30', '2:': '02:00',
                                '::': '', '14:0015:00': '14:30', '09:0009:30': '09:15', '06:0007:20': '06:40', ':12:00': '12:00', '11:0011:30': '11:15', '19:0020:00': '19:30',
                                    '11:115': '11:15'})

In [144]:
mask5 = data['Time'] != ''

data.loc[mask5, 'Time'] = pd.to_datetime(data.loc[mask5, 'Time'], format='%H:%M').dt.time

In [145]:
data['Time'].replace("", "unknown", inplace=True)

In [146]:
data.Species.value_counts().head(60)

White shark                                           162
Shark involvement prior to death was not confirmed    105
Invalid                                               101
Shark involvement not confirmed                        87
Tiger shark                                            70
Shark involvement prior to death unconfirmed           68
Bull shark                                             47
4' shark                                               40
6' shark                                               39
Questionable incident                                  35
1.8 m [6'] shark                                       33
Questionable                                           32
1.5 m [5'] shark                                       32
3' shark                                               26
1.2 m [4'] shark                                       26
5' shark                                               26
2 m shark                                              25
4' to 5' shark

In [147]:
data.Species.isnull().sum()

2743

In [148]:
data.Species.fillna('unknown', inplace=True)

In [149]:
data['Species'] = np.where(data['Species'].str.contains('hite'), 'White shark', data['Species'])

data['Species'] = np.where(data['Species'].str.contains('iger'), 'Tiger shark', data['Species'])

data['Species'] = np.where(data['Species'].str.contains('ull'), 'Bull shark', data['Species'])

data['Species'] = np.where(data['Species'].str.contains('lue'), 'Blue shark', data['Species'])

data['Species'] = np.where(data['Species'].str.contains('urse'), 'Nurse shark', data['Species'])

data['Species'] = np.where(data['Species'].str.contains('ako'), 'Mako shark', data['Species'])

data['Species'] = np.where(data['Species'].str.contains('hammer'), 'Hammerhead shark', data['Species'])

In [150]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6169 entries, 0 to 6168
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Day                     6169 non-null   int32 
 1   Month                   6169 non-null   int32 
 2   Year                    6169 non-null   int32 
 3   Type                    6169 non-null   object
 4   Country                 6169 non-null   object
 5   Area                    6169 non-null   object
 6   Location                6169 non-null   object
 7   Activity                6169 non-null   object
 8   Name                    6169 non-null   object
 9   Sex                     6169 non-null   object
 10  Age                     6169 non-null   object
 11  Injury                  6169 non-null   object
 12  Fatal                   6169 non-null   object
 13  Time                    6169 non-null   object
 14  Species                 6169 non-null   object
 15  Inve

In [151]:
data.Investigator_or_Source.value_counts().head(60)

C. Moore, GSAF                                                                                        101
C. Creswell, GSAF                                                                                      92
S. Petersohn, GSAF                                                                                     82
R. Collier                                                                                             55
R. Collier, GSAF                                                                                       48
T. Peake, GSAF                                                                                         48
M. Levine, GSAF                                                                                        45
A. Gifford, GSAF                                                                                       28
C. Moore. GSAF                                                                                         27
B. Myatt, GSAF                                

In [152]:
data['Investigator_or_Source'].fillna('unknown', inplace=True)

In [153]:
data.pdf.value_counts().tail(60)

1999.03.05-Knutson.pdf                1
1999.07.29-Capri.pdf                  1
1999.02.26.R-BirdDog.pdf              1
1999.02.23-ScottsHead.pdf             1
1999.03.18.a-Davoodabai.pdf           1
1999.02.03-KennyBurns.pdf             1
1999.01.07-NZ-inflatable.pdf          1
1999.01.03.R-Turcotte.pdf             1
1999.01.03-ReunionIsland.pdf          1
1999.00.00.b-Lootz.pdf                1
1999.00.00.a-NV-SandridgeBeach.pdf    1
2000.09.15-Smith.pdf                  1
2000.09.16.a-NV-IsleOfPalms.pdf       1
2000.03.30-Rojcevic.pdf               1
2000.10.06.b-White.pdf                1
2000.11.10-Skeie.pdf                  1
2000.11.06.b-Avery.pdf                1
2000.11.06.a-KenCrew.pdf              1
2000.11.04-Stewman.pdf                1
2000.10.29-Kelly.pdf                  1
2000.10.20-Licamele.pdf               1
2000.10.18-Musselwhite.pdf            1
2000.10.14-Payne.pdf                  1
2000.10.09-Kraskiecwicz.pdf           1
2000.10.06.a-Holley.pdf               1


In [154]:
data.loc[data['href_formula'].isnull()]

Day  Month  Year        Type    Country             Area    Location  \
3214   19      1  1975  Unprovoked  AUSTRALIA  South Australia  Coffin Bay   

     Activity             Name Sex Age Injury Fatal     Time  Species  \
3214  Surfing  David Barrowman   M  17  FATAL     Y  unknown  unknown   

                                      Investigator_or_Source  \
3214  J. West; Adelaide Advertiser, 1/20/1975; P. Kemp, GSAF   

                           pdf href_formula  \
3214  1975.01.19-Barrowman.pdf          NaN   

                                                                                href  \
3214  http://sharkattackfile.net/spreadsheets/pdf_directory/1975.01.19-Barrowman.pdf   

     Case_Number_1 Case_Number_2  original_order Unnamed_22 Unnamed_23  
3214    1975.01.19    1975.01.19            3059        NaN        NaN

In [155]:
data.href_formula.iloc[3214] = data.href.iloc[3214]

In [156]:
data.href_formula.iloc[3214]

'http://sharkattackfile.net/spreadsheets/pdf_directory/1975.01.19-Barrowman.pdf'

In [157]:
data.iloc[3214]

Day                                                                                                   19
Month                                                                                                  1
Year                                                                                                1975
Type                                                                                          Unprovoked
Country                                                                                        AUSTRALIA
Area                                                                                     South Australia
Location                                                                                      Coffin Bay
Activity                                                                                         Surfing
Name                                                                                     David Barrowman
Sex                                                    

In [158]:
data.rename(columns={'Case_Number_1': 'Case_Number', 'Case_Number_2': 'Date', 'Unnamed_22': 'no_data_1', 'Unnamed_23': 'no_data_2'}, inplace=True)

In [159]:
len(data)

6169

In [160]:
data['Case_Number'] = [len(data) - i for i in range(len(data))]

In [161]:
data.head(60)

Day  Month  Year        Type                  Country  \
0    25      6  2018     Boating                      USA   
1    23      2  2018  Unprovoked                AUSTRALIA   
2    15      4  2018  Unprovoked                AUSTRALIA   
3    14      4  2018  Unprovoked                  BAHAMAS   
4    10      1  2018     Invalid                   BRAZIL   
5     9      4  2018  Unprovoked            NEW CALEDONIA   
6     5      4  2018  Unprovoked                  BAHAMAS   
7     3      4  2018  Unprovoked             SOUTH AFRICA   
8    31      3  2018  Unprovoked                      USA   
9    14      3  2018  Unprovoked                AUSTRALIA   
10    9      3  2018  Unprovoked                AUSTRALIA   
11    9      3  2018  Unprovoked                AUSTRALIA   
12   24      2  2018  Unprovoked             SOUTH AFRICA   
13   17      2  2018    Provoked                AUSTRALIA   
14   15      4  2018  Unprovoked                   BRAZIL   
15   15      2  2018  Unprovoked                AUSTRALIA   
16   11      2  2018     Invalid                   BRAZIL   
17    3      2  2018  Unprovoked                  ECUADOR   
18    1      2  2018     Invalid                AUSTRALIA   
19   28      1  2018  Unprovoked                AUSTRALIA   
20   21      1  2018  Unprovoked            NEW CALEDONIA   
21   14      1  2018  Unprovoked                AUSTRALIA   
22   13      1  2018  Unprovoked                AUSTRALIA   
23   12      1  2018  Unprovoked                   BRAZIL   
24    5      1  2018  Unprovoked                AUSTRALIA   
25   18      6  2018  Unprovoked                      USA   
26   15      4  2018  Unprovoked                AUSTRALIA   
27   14      2  2018     Invalid                AUSTRALIA   
28   15      4  2018  Unprovoked                 THAILAND   
29   13      5  2018     Invalid           UNITED KINGDOM   
30   19      4  2018  Unprovoked                      USA   
31    9      6  2018     Invalid                      USA   
32    8      6  2018  Unprovoked                AUSTRALIA   
33    4      6  2018    Provoked                   MEXICO   
34    3      6  2018  Unprovoked                AUSTRALIA   
35    3      6  2018  Unprovoked                   BRAZIL   
36   26      5  2018  Unprovoked                      USA   
37   26      5  2018  Unprovoked                      USA   
38   24      5  2018    Provoked                AUSTRALIA   
39   21      5  2018  Unprovoked                      USA   
40   13      5  2018  Unprovoked                      USA   
41   27      5  2018  Unprovoked                      USA   
42   27      5  2018    Provoked                AUSTRALIA   
43   28      4  2018  Unprovoked                AUSTRALIA   
44   22      4  2018  Unprovoked             SOUTH AFRICA   
45   23      4  2018  Unprovoked                 MALDIVES   
46   24      4  2018  Unprovoked                AUSTRALIA   
47   25      4  2018  Unprovoked                AUSTRALIA   
48   28      4  2018  Unprovoked               COSTA RICA   
49   30      1  2018  Unprovoked                 THAILAND   
50   12      5  2018  Unprovoked             SOUTH AFRICA   
51   27      5  2017     Invalid                AUSTRALIA   
52   29      4  2017  Unprovoked             SOUTH AFRICA   
53   12      5  2017  Unprovoked     UNITED ARAB EMIRATES   
54    6      1  2017    Provoked                AUSTRALIA   
55    6      5  2017  Unprovoked                   MEXICO   
56    3      5  2017     Invalid                      USA   
57   29      4  2017  Unprovoked                      USA   
58   22      4  2017  Unprovoked  UNITED KINGDOM OVERSEAS   
59   29      4  2017  Unprovoked                      USA   

                     Area                                           Location  \
0              California                        Oceanside, San Diego County   
1         New South Wales                 Little Congwong Beach, La Perouse    
2       Western Australia                 

In [162]:
data.no_data_1 = 0

In [163]:
data['Date'] = data['Year'].astype(str) + '-' + data['Month'].astype(str) + '-' + data['Day'].astype(str)

In [164]:
data.no_data_2 = 0

In [165]:
data.shape

(6169, 24)

In [170]:
((data == 'unknown').sum(axis=1) > 11).unique()

array([False])

In [171]:
data.head()

Day  Month  Year        Type    Country               Area  \
0   25      6  2018     Boating        USA         California   
1   23      2  2018  Unprovoked  AUSTRALIA    New South Wales   
2   15      4  2018  Unprovoked  AUSTRALIA  Western Australia   
3   14      4  2018  Unprovoked    BAHAMAS     New Providence   
4   10      1  2018     Invalid     BRAZIL            Alagoas   

                             Location  Activity                  Name Sex  \
0         Oceanside, San Diego County   unknown           Julie Wolfe   F   
1  Little Congwong Beach, La Perouse   Swimming        Anna Shurapey    F   
2   Cobblestones, Margaret River Area   Surfing  Alejandro Travaglini   M   
3                       Nirvana Beach   Surfing           Bruce Rowan   M   
4           Praia de Sauaçuhy, Maceió   Fishing            Josias Paz   M   

       Age Injury Fatal      Time                          Species  \
0       57  FATAL     N  18:00:00                      White shark   
1       55  FATAL     N  19:00:00                      White shark   
2       37  FATAL     N  08:00:00                          unknown   
3  unknown  FATAL     N  09:30:00                      Tiger shark   
4       56  FATAL     N   unknown  Shark involvement not confirmed   

            Investigator_or_Source                          pdf  \
0                 R. Collier, GSAF         2018.06.25-Wolfe.pdf   
1                   B. Myatt, GSAF      2018.02.23-Shurapey.pdf   
2                    B.Myatt, GSAF  2018.04.15.a-Travaglini.pdf   
3                      Tribune242,         2018.04.14-Rowan.pff   
4  K. McMurray, TrackingSharks.com         2018.04.10.R-Paz.pdf   

                                                                        href_formula  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

                                                                                href  \
0         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.06.25-Wolfe.pdf   
1      http://sharkattackfile.net/spreadsheets/pdf_directory/2018.02.23-Shurapey.pdf   
2  http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.15.a-Travaglini.pdf   
3         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.14-Rowan.pff   
4         http://sharkattackfile.net/spreadsheets/pdf_directory/2018.04.10.R-Paz.pdf   

   Case_Number       Date  original_order  no_data_1  no_data_2  
0         6169  2018-6-25            6303          0          0  
1         6168  2018-2-23            6263          0          0  
2         6167  2018-4-15            6274          0          0  
3         6166  2018-4-14            6273          0          0  
4         6165  2018-1-10            6272          0          0

In [173]:
data['Time'].value_counts()

unknown     3372
16:00:00     330
09:00:00     187
11:00:00     130
12:00:00     125
            ... 
13:42:00       1
13:19:00       1
12:39:00       1
17:46:00       1
01:50:00       1
Name: Time, Length: 250, dtype: int64